### PR TITLE
Sockets- Tatiana

### PR DESCRIPTION
**Pull request by tatsqui**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | Designing everything myself and sticking with a plan was the most challenging part. In addition to figuring out how to pass information around between the different classes - I struggled with visualizing this well and I plan on working on this more.  
What was a design decision you made that changed over time over the project? | How and where to use date-range class. I started off with no class, then I created a class, then I deleted that class, ultimately to bring it back into play. I decided to bring it back because it helped to DRY up the code a bit. 
What was a concept you gained clarity on, or a learning that you'd like to share? | Honestly, I think the thing I did better than anything was learning how to write tests, and how writing a test before writing code can be very useful in planning towards the next steps in writing code. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | When trying to book a room, a nominal test is booking a room that has no date-range conflicts - we expect this to pass. 
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | When a room is being booked for a date that ends on another reservation's start date for the same room. 
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I did a lot better this time around, and I like the practice of psuedocode and test writing which helped me get to produce something that is useable.

### CODE REVIEW

File: lib/booker.rb [hunk 1]
```diff
@@ -0,0 +1,95 @@
+require "time"
+require "date"
+
+require_relative "room"
+require_relative "reserve"
+require_relative "date_range"
+require_relative "blocks"
+
+class RoomBooker < Date
```
Comment (dHelmgren) [inline lib/booker.rb]: I don't know why the `RoomBooker` is extending `Date` here. This implies that the `RoomBooker` is a kind of `Date`, but it seems to be a manager class.

File: lib/blocks.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  attr_reader :blocked_rooms, :total_cost
+  ROOM_BLOCK = 5
+
+  def initialize(check_in:, check_out:, blocked_rooms:, discount:)
+    # unless id.instance_of?(Integer) && id > 0 && id <= 20
+    #   raise ArgumentError, "ID must be a positive number, given #{id}..."
+    # end
+
+    @check_in = Date.parse(check_in)
+    @check_out = Date.parse(check_out)
```
Comment (dHelmgren) [inline lib/blocks.rb]: You can assume that people pass you pre-parsed Date Objects, you don't have to do all this internal parsing yourself.

File: lib/booker.rb [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  end
+
+  # working on blocked rooms
+
+  def reserve_block(check_in:, check_out:, rooms_needed:)
+    raise ArgumentError if rooms_needed > 5
+    block_of_rooms = []
+
+    rooms_needed.times do
+      room = find_available_room(check_in: check_in, check_out: check_out)
```
Comment (dHelmgren) [inline lib/booker.rb]: What if you run out of rooms partway thru?

File: lib/blocks.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+require "date"
+
+class BlockParty
```
Comment (dHelmgren) [inline lib/blocks.rb]: Ha

File: lib/reserve.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
+require "date"
+require_relative "room"
+require "awesome_print"
+
+class Reservation
```
Comment (dHelmgren) [inline lib/reserve.rb]: Try to keep your file/class names in parity, otherwise it's hard to know which files to open.

File: lib/room.rb [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    @bookings << (Date.parse(check_in)..Date.parse(check_out))
+  end
+
+  def room_available?(check_in:, check_out:)
+    check_in = Date.parse(check_in)
+    check_out = Date.parse(check_out)
+    @bookings.each do |booked|
+      if booked == nil
+        return true
+      end
```
Comment (dHelmgren) [inline lib/room.rb]: less clunky way of doing this: just return true at the top if this array is empty.

File: lib/booker.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  attr_reader :reservations, :rooms, :get_available_rooms
+
+  def initialize(rooms:)
+    @rooms = rooms
+    @reservations = []
+    @blocked_reservations = []
+  end
+
+  def book_reservation(check_in:, check_out:)
+    date_request = DateRange.new(check_in: check_in, check_out: check_out)
```
Comment (dHelmgren) [inline lib/booker.rb]: It looks like you just kind of create and then throw away `DateRange` objects. Are they really needed ?

File: lib/room.rb [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+  end
+
+  def booked_on(check_in:, check_out:)
+    @bookings << (Date.parse(check_in)..Date.parse(check_out))
+  end
+
+  def room_available?(check_in:, check_out:)
+    check_in = Date.parse(check_in)
+    check_out = Date.parse(check_out)
+    @bookings.each do |booked|
```
Comment (dHelmgren) [inline lib/room.rb]: I get really confused by what `booked` means in this loop, I think because it's a verb and not a noun. I'd have gone with `booking`, or `old_booking` for clarity.

File: lib/room.rb [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+        return true
+      end
+      if (check_in..check_out).include?(booked)
+        return false
+      end
+      if check_out == booked.first || check_in == booked.last
+        return true
+      elsif booked.include?(check_in) || booked.include?(check_out) || check_in <= booked.first && check_out >= booked.last
+        return false
+      end
```
Comment (dHelmgren) [inline lib/room.rb]: So, from lines 38 to 42, I'd just use the `elsif` part of this and then let the thing fall thru to a true.

File: lib/booker.rb [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    end
+    if open_rooms.length == 0
+      raise ArgumentError, "No rooms are available for the provided date range."
+    end
+    return open_rooms
+  end
+
+  # working on blocked rooms
+
+  def reserve_block(check_in:, check_out:, rooms_needed:)
```
Comment (dHelmgren) [inline lib/booker.rb]: It seems like you can reserve a block, but you can't then reserve a room that's in that block.

File: lib/date_range.rb [hunk 11]
```diff
@@ -0,0 +1,31 @@
+
+class DateRange
```
Comment (dHelmgren) [inline lib/date_range.rb:2]: I don't think this class is really giving you anything valuable.

File: lib/booker.rb [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+  end
+
+  def find_available_room(check_in:, check_out:)
+    found_room = rooms.select { |room| room.room_available?(check_in: check_in, check_out: check_out) }
+    return found_room[0]
+  end
+
+  def date_query(date:)
+    find_dates = Date.parse(date)
+    find_by_dates = reservations.select { |reserved| reserved.dates_booked.include?(find_dates) == true }
```
Comment (dHelmgren) [inline lib/booker.rb]: I don't like how these are named. `find_dates` implies more than one, and the query accepts one date.

`find_by_dates` sounds like a method to me, it implies a verb is happening. Better alternatives: `found_reservations` or `booking_list`

File: lib/booker.rb [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+  def find_cost(id:)
+    @reservations.each do |res|
+      if res.id == id
+        return "#{res.total_cost}"
+      end
+    end
+    return nil
+  end
+
+  def get_available_rooms(check_in:, check_out:)
```
Comment (dHelmgren) [inline lib/booker.rb]: Why is there a `get_available_rooms` and a `find_available_room`? These seem to be trying to do the same job.

File: Hotel.rb [hunk 14]
```diff
@@ -0,0 +1,22 @@
+require 'csv'
```
Comment (dHelmgren) [inline Hotel.rb:1]: Why did this file make it to the submission? It doesn't seem to be doing anything.

File: spec/booker_spec.rb [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+    end
+
+    it "will not book a reservation for a check-in date that occurs after check-out date" do
+      expect {
+        @second_reservation = @second_hotel.book_reservation(check_in: "September 15th 2023", check_out: "September 8th 2023")
+      }.must_raise ArgumentError
+    end
+  end
+
+  describe "book reservation" do
```
Comment (dHelmgren) [inline spec/booker_spec.rb]: I think you have a describe block named the same thing above. Why do you have both?

File: lib/room.rb [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+  def room_available?(check_in:, check_out:)
+    check_in = Date.parse(check_in)
+    check_out = Date.parse(check_out)
+    @bookings.each do |booked|
+      if booked == nil
+        return true
+      end
+      if (check_in..check_out).include?(booked)
+        return false
```
Comment (dHelmgren) [inline lib/room.rb]: According to your cov files, lines 33 and 36 don't get hit! This could either be because you didn't write a test for them, or because they cannot be hit.

File: lib/booker.rb [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+      room = find_available_room(check_in: check_in, check_out: check_out)
+      room.booked_on(check_in: check_in, check_out: check_out)
+      block_of_rooms << room
+    end
+
+    if block_of_rooms.length < rooms_needed
+      raise ArgumentError, "We cannot book this block reservation due to insufficient room availability"
+    elsif block_of_rooms.length > 5
+      raise ArgumentError, "We cannot block more than 5 rooms per party."
+    end
```
Comment (dHelmgren) [inline lib/booker.rb]: This block is really good, but it looks like you didn't write tests that triggered this code!

**Comment by dHelmgren** [comment]
# Hotel
## What We're Looking For

Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | yes
Answer comprehension questions | yes
**Design** |
Each class is responsible for a single piece of the program | Yes
Classes are loosely coupled | Yes
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | Your tests didn't hit some of this, see comments
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | Good
**Wave 3** |
Create a block of rooms | Yes
Check if a block has rooms | Not really? 
Reserve a room from a block | The former is due to, I think, a misunderstanding of the instructions. We were looking to be able to book rooms that were part of a block as a separate action, while you treat a block as a reservation in its own right.
Test coverage | Some missed tests, see comments.
**Fundamentals** |
Names variables, classes and modules appropriately | Not so much, see comments
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | Mostly good, but see comments
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | yes
Appropriately writes and utilizes classes | No, see comments
Appropriately utilizes modules as a namespace | No, you have a module, but you aren't using it as a namespace.
**Wrap Up** |
There is a refactors.txt file | no
The file provides a roadmap to future changes | no
**Additional Feedback** | There's a lot of good stuff going on here, but it's put together in a way that makes it harder to read and use than it needs to be. With a project like this, really going back and cleaning things up after you get them running is important. It feels to me like you wanted to get the whole of the project done, but that some of the requirements upset the apple cart as you went, and you never circled back to do your refactoring.
